datafusion/src/physical_plan/functions.rs

```diff
@@ -543,6 +553,28 @@ macro_rules! invoke_if_unicode_expressions_feature_flag {
     };
 }
 
+///Function volatility determines when a function can be inlined or in the future have evaluations elided when the arguments are the same
+///Immutable - a pure function which always remains the same
+///Stable - Maybe not required?. Used for functions which can modify the db but are stable for multiple calls in statement
+///Volatile - Functions where output can vary for each call.
+///For more information see https://www.postgresql.org/docs/current/xfunc-volatility.html
+#[derive(Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
+#[repr(u32)]
+pub enum FunctionVolatility {
+    ///Immutable - a pure function which always remains the same
+    Immutable = 0,
+    ///Stable - Maybe not required?. Used for functions which can modify the db but are stable for multiple calls in statement is this relevant for datafusion as in mem?
+    Stable = 1,
+    ///Volatile - Functions where output can vary for each call.
+    Volatile = 2,
```

Comment:
```suggestion
    /// A pure function - one whose always returns the same given inputs and has no side-effects. (E.g. `SQRT`)
    Immutable,
    /// A function whose output can vary for each call. (E.g. `CURRENT_DATE`)
    Volatile,
```

Reply:
FWIW `now()` is an example of a stable function I believe

The distinction is that within a query, all instances of `now()` in the same query evaluate to the same value, though for different queries they may evaluate differently. Volatile functions such as `rand()` really can and do produce a different value for each invocation even within the same query.

So from DataFusion's perspective, we should be able to fold stable and immutable functions (not volatile)

Reply:
That's a good point @alamb , I agree and I think we should keep the Stable variant.